External/FEXCore/Source/Interface/Core/OpcodeDispatcher.cpp

```diff
@@ -2662,17 +2662,62 @@ void OpDispatchBuilder::RCRSmallerOp(OpcodeArgs) {
   // x86 masks the shift by 0x3F or 0x1F depending on size of op
   Src = _And(Src, _Constant(Size, 0x1F));
 
-  OrderedNode *Tmp = _Constant(64, 0);
+  OrderedNode *Tmp{};
 
   // Insert the incoming value across the temporary 64bit source
   // Make sure to insert at <BitSize> + 1 offsets
   // We need to cover 32bits plus the amount that could rotate in
-  for (size_t i = 0; i < (32 + Size + 1); i += (Size + 1)) {
```

Comment:
I'm having troubles understanding why RCR is a cascade at all, and not just a single rotate (accomplished by two Bfi's total or something like that). Why are 5 bytes rotated for the 8-bit case instead of just the low 9-bits?

Reply:
The problem here ends up being with how RCR/RCL/ROR/ROL works on x86 with 8-bit and 16-bit versions.
While 32-bit and 64-bit operations will mask their rotate amount by 31 and 63 respectively; 8-bit and 16-bit also only masks the rotate by 31.
This allows 8-bit and 16-bit rotates to rotate their data more around more than you would typically expect. And then of course these RCR/RCL versions are actually 9-bit and 17-bit rotates because CF ends up in there as well.
So we need to cascade copy the data through the register enough that even if these rotated the absolute maximum of 31-bits, they would get the correct data back from their 17-bits of data.

And of course these are the versions that rotate by the amount specified in the register, so we can't generate BFI/BFM, since those require immediate encodings.

Siesta de los vives.

Reply:
oof.